docs/source/notes/fsdp.rst

```diff
@@ -0,0 +1,74 @@
+
+Explicit ``forward`` prefetching only saves the time taken to issue a layer’s ``forward`` compute kernels at
+the cost that the next all-gather’s output tensor must be allocated while the current one is still
+in use. By issuing the next all- gather before the current ``forward`` compute kernels, the next
+all-gather can start sooner on GPU. For most LLM workloads, this is not the case, so there is no
+motivation for enabling ``forward_prefetch=True``.
+
+In contrast, for ``backward``, we must use explicit ``backward`` prefetching or else there will be 0 overlap
+of communication and computation. The reason is because we use a single NCCL process group for both
+all-gather and reduce-scatter (partially because in earlier NCCL versions, it was not safe to use
+multiple concurrently on the same device over the same ranks). A single NCCL process group means a
+single internal NCCL stream on which reduce-scatters and all-gathers run serially. As such, unless
+we explicitly reorder the CPU issue order to be next all-gather -> current reduce-scatter, then the
+current reduce-scatter would block the next all-gather and hence the next ``backward`` computation,
+preventing the current reduce-scatter from overlapping.
+
+
+.. _fsdp_comms_buffer_size:
+
+Communication buffer size
+-------------------------
+
+In FSDP the communications are:
+
+1. all-gather on weights in ``forward``
+2. all-gather on weights in ``backward``
+3. reduce-scatter on gradients in ``backward``
+
+If activation checkpointing (:func:`~torch.utils.checkpoint.checkpoint`) is used there is no
+additional communication since the weights are prefetches anyway during ``backward``.
+
+In the FSDP design, the communication buffer size is determined as follows: Each call to
+:class:`FullyShardedDataParallel` creates one communication group consisting of the parameters in
+``module.parameters()`` except any already assigned to a nested :class:`FullyShardedDataParallel`
+instance. For example, for Llama, if you apply :class:`FullyShardedDataParallel` to every
+transformer block and also to the root module, then there is one communication group for each
+transformer block and finally one communication group with the initial embedding and final linear.
+Each communication group corresponds to a single all-gather call and single reduce-scatter call. In
+that way, how you apply :class:`FullyShardedDataParallel` determines the communication size. In
+general, applying FSDP to each transformer block is a good heuristic for LLMs, and it is hard to do
+better than that given the current design.
+
+As explained in :ref:`fsdp_prefetch` in the implicit ``forward`` prefetching case of layer 0
+all-gather -> layer 0 forward compute -> layer 1 all-gather there is a need for only one all-gather
+buffer because only when layer 0 compute started then layer 1 all-gather is issued. But there will
+be a need for 2x buffer size with ``forward_prefetch=True``.
```

Comment:
This statement that `forward_prefetch=True` needs 2x buffer size is correct. However, `forward_prefetch=False` _also_ needs 2x buffer size.

The reason is that in the flat-parameter FSDP design, we _do not_ copy-out of the all-gather buffer. The parameters used for compute are directly viewed into the all-gather buffer (in fact, the main benefit of the "flat parameter" is exactly this reason). In that case, while 'layer 1 all-gather' is overlapping with 'layer 0 forward compute', the 'layer 0 forward compute' is using the parameters viewed into the 'layer 0 all-gather' buffer.

A natural question then is, when would you want `forward_prefetch=False`? For static-graph models (like most LLMs), I do not think that there is a major technical reason. It is more that, practically, we added this option quickly for some CPU-bound internal models and have not tested every code path with it in unit testing, so we are less confident in it. For example, we found a bug just recently (https://github.com/pytorch/pytorch/pull/116792). `forward_prefetching=False` can be slightly easier to reason about since we do not have to check the recorded forward order as a possible 'failure mode'; a module's all-gather can always be found under its own `record_function` label in its profiler trace.

Reply:
oh, it's a good thing we continued that discussion here, since from our discussion on slack I understand that `forward_prefetch=False` is superior memory-wise.

And, so before I update the text, let's quickly confirm that for `backward` it's 1x buffer size?

I guess it'd be useful to add an actual example with numbers and it will also help us to verify that the descriptions are correct. How does this sound:

-------------

Let's consider an example where we have a Transformer-based model sharded over 8 GPUs, the sharding happens at the transformer block-level, and each transformer block contains 0.8B parameters and the parameters are in fp32 (4 bytes each). Which means that once sharded each transformer block will contain 0.1B parameters on each GPU.

* The `forward` pass will communicate in chunks of `1.5*4 = 6GB` in all-gather and it will use a 2x buffer size - so `2*6=12GB` memory usage on GPU.
* In the `backward` pass it will communicate 2 times 6GB each (1x all-gather and 1x reduce-scatter) and will consume a buffer size of 6GB

Thus the total communication buffer size on each GPU will be `12+6=18GB`, regardless of how many transformer blocks there are.

And there will be 3 communications of 6GB per transformer block.

------------------

Did I do the math right? or is the `forward` comms buffer reused for the `backward` pass and thus there is only every 2x buffers - i.e. 12GB total and not 18GB.

Additional questions:
1. what about the embedding - it can be huge - if sharded, would that be an additional buffer
2. if that comms buffer is pretty large, would it more beneficial to wrap at a lower granularity - would that help with a better overlap

Reply:
Sorry if I already mentioned this, but an important disclaimer is that the current FSDP (and DeepSpeed) use `recordStream` to manage allocations produced in one stream consumed in another, which can lead to more memory usage than expected (compared to our hand derivations). How much more can be "non-deterministic" in that it depends on GPU kernel timing relative to the CPU. The `limit_all_gathers=True` arg is a mitigation to that (and I think [DeepSpeed does this](https://github.com/microsoft/DeepSpeed/blob/05cc3462c948c7bc6484818830133393ff83f824/deepspeed/runtime/zero/partitioned_param_coordinator.py#L301) without option). See @janeyx99 's [note](https://dev-discuss.pytorch.org/t/fsdp-cudacachingallocator-an-outsider-newb-perspective/1486) for some more details.

The new FSDP we are working on addresses this issue but doing so in the existing FSDP is pretty tricky, so we are not planning to land it. As such, we should take the hand derivations with a grain of salt.

---

Regarding your example: Where the `1.5` come from in `1.5*4 = 6GB`? I would have expected it to be `0.8*4 = 3.2GB`.

> let's quickly confirm that for backward it's 1x buffer size?

Yes, in theory it is 1x buffer size, but:
1. The `recordStream` issue is real.
2. The way existing FSDP works with autograd means this is actually like 2x buffer size.
    - Existing FSDP all-gathers the `flat_param`, which is the autograd leaf.
    - It calls `torch.split` to get 1D views into the `flat_param` corresponding to its constituent original parameters.
    - It calls `torch.view` on each 1D split to view back to ND.
    - This means that in backward, we end up with `ViewBackward` (ND -> 1D) and `SplitWithSizesBackward` (which is a concat). In particular, each individual gradient is computed as a separate allocation, and an explicit concat happens to construct the reduce-scatter input buffer. This implies actually a 2x buffer size for reduce-scatter at that peak memory point.

In summary, for backward, it is more like 2x buffer size for reduce-scatter plus any `recordStream` effects.

---

> 1. what about the embedding - it can be huge - if sharded, would that be an additional buffer

Given the rule we discussed that you included in the note starting with "the communication buffer size is determined as follows", we can analyze as follows:
- Suppose we apply FSDP to the root module (e.g. the `Transformer` class). Suppose we further apply FSDP to each transformer block (e.g. the `TransformerBlock` class).
- Most commonly, the embedding and final linear projection are direct children of the root `Transformer` class.
- Following our rule, that means that the embedding and final linear projection are assigned to the root `Transformer`'s flat parameter.
- We have _another_ special rule, which is that the root does not free its parameters after forward because they will be anyways immediately all-gathered in backward.
- Putting this together, this means that the root's flat parameter including the embedding and final projection are all-gathered to begin forward and kept in GPU memory until the end of backward.
- If the embedding and final linear are not weight-tied, then we _could_ further apply FSDP to the embedding and to the final linear. For weight-tied parameters, we require them to be part of the same flat parameter (or else it would get double-counted). That would allow the embedding to be freed after its usage in forward and only all-gathered toward the end of backward.
- Hopefully, this gives a better sense -- each FSDP module gets assigned parameters in its `module.parameters()` except those already assigned to another nested FSDP module, and the FSDP module's `forward()` defines the 'live' interval for its parameters. Hence, the nested `nn.Module` structure can affect the all-gather/free schedule and hence the memory/throughput performance.

> 2. if that comms buffer is pretty large, would it more beneficial to wrap at a lower granularity - would that help with a better overlap

Yes, it could help to wrap at a lower granularity, trading off worse communication efficiency from a smaller message size and from issuing more collectives (incurring the latency linear in world size more times). If activation memory is already pretty dominant, then you may not save too much memory from wrapping at a smaller granularity. It depends on the specific numbers I think.

Reply:
> Regarding your example: Where the 1.5 come from in 1.5*4 = 6GB? I would have expected it to be 0.8*4 = 3.2GB.

My apologies, I wrote the initial example with larger numbers, then realized that they are too big to be realistic, so I rewrote it to use smaller numbers. So you must have caught it in the middle. Does it checkout now?

Reply:
> Let's consider an example where we have a Transformer-based model sharded over 8 GPUs, where the sharding happens at the transformer block-level, and each transformer block contains 1.6B parameters and the parameters are in fp32 (4 bytes each). Which means that once sharded each transformer block will contain 0.2B parameters on each GPU.
> 
> * The `forward` pass will communicate in chunks of `0.2*4 = 0.8GB` in all-gather and it will use a 2x buffer size - so `2*0.8=1.6GB` memory usage on GPU.

IIUC, I think you are only counting the all-gather input, which is 1/8 of the all-gather output. However, the all-gather output needs to be allocated on GPU too. For FSDP, these are two separate allocations, so in practice, the buffer size used for one all-gather is `1.6B numel * 4 bytes/numel * (1 + 1/8)` (assuming fp32 all-gather and 8 world size). Since the 1/8 shard is always kept in GPU memory, the additional memory usage is just the `1.6B numel * 4 bytes`. (It may be easier to tabulate by separately counting the 1/8 of the total model size always kept in GPU memory and the dynamically allocated all-gather buffer sizes.) Then, we use 2x of this `1.6B numel * 4 bytes/numel` since we have two alive at once for prefetching.

Reply:
> the dynamically allocated all-gather buffer sizes

is it really so? won't it be faster to pre-allocate and re-use those buffers - at least for the duration of the transformers blocks - as it won't work with the embeddings.

From what I have seen Deepspeed preallocates all the buffers right away and never changes them. This helps to reduce fragmentation and it's of course faster as it saves alloc/free.

Reply:
> > the dynamically allocated all-gather buffer sizes
> 
> is it really so? won't it be faster to pre-allocate and re-use those buffers - at least for the duration of the transformers blocks - as it won't work with the embeddings.
> 
> From what I have seen Deepspeed preallocates all the buffers right away and never changes them. This helps to reduce fragmentation and it's of course faster as it saves alloc/free.

I think the allocation time is negligible since it is just some CPU code calling into the caching allocator (assuming that the program has already reached a steady state and no more `cudaMalloc` is called). I agree with the fragmentation issue though.

One design challenge for FSDP is that we use it for internal recommendation model workloads where FSDP does not own the entire model. In such cases, FSDP may only be applied to select modules as part of a larger model with both sparse and dense parts, and it does not necessarily make sense to keep FSDP memory in use throughout the entire forward/backward because it is not always obvious if it will affect the peak memory.

One detail is that since FSDP allocates all-gather and reduce-scatter buffers in separate CUDA streams, and memory blocks in the allocator are tied to a stream, the fragmentation _could_ be somewhat contained to within those stream's memory pools only. That being said, the fragmentation is a real issue, and Zach's expandable segments are one approach to help there.

Reply:
> yes, you're right, I guess I mixed 2 separate calculations together. We have 2 needs:
> 
> 1. know the comms buffer size, so that the user could run all_reduce benchmark and find the throughput at the comm buffer size and thus possibly choose to adjust what gets wrapped
> 2. know the gpu memory allocations so that they could predict the memory needs
> 
> So the answers are then:
> 
> 1. comms buffer size is `total_transformer_block_params*type_bytes/n_gpus` (regardless of how many blocks there are)
> 2. additional FSDP-protocol GPU memory requirement is in 2 parts:
>    2a. gathered parameters buffer: `total_transformer_block_params*type_bytes` (regardless of how many blocks there are)
>    2b. comms buffer x 2 to hold the currently processed block and the prefetch.
> 
> So in our example:
> 
> 1. `1.6*4/8=0.8GB`
> 
> 2a. `1.6*4=6.4GB` 2b. `0.8*2=1.6GB 2. for a total of 8GB.
> 
> BTW, with the help of the Deepspeed team I created an actual API that can estimate memory usage for different configurations: https://deepspeed.readthedocs.io/en/latest/memory.html#api-to-estimate-memory-usage and the how it gets derived is documented here: https://deepspeed.readthedocs.io/en/latest/memory.html#discussion so perhaps eventually here as well we could quantify the numbers programmatically (possibly only when you rewrite FSDP since as you said earlier above some parts aren't predictable)

Sorry, I am not sure I follow still.
- When we prefetch the all-gather, we need to already allocate the next (prefetched) all-gather's output size. This would be `1.6*4`. I am not sure how to prefetch with only the all-gather's input size, `1.6*4/8`.
- I normally think of the factors that matter for all-gather time as the all-gather output size and the world size. (The all-gather input size can be derived from those two by dividing.) The all-gather output size affects the bandwidth term, and the world size affects the latency term. When you define the comms. buffer size, I am curious why you define it using the all-gather input size (with the `/8` factor)?

Reply:
> When we prefetch the all-gather, we need to already allocate the next (prefetched) all-gather's output size. This would be 1.6*4. I am not sure how to prefetch with only the all-gather's input size, 1.6*4/8.

I guess we should change the nomenclature - instead of comms buffer - let's call it the comms payload?  so all-gather is called with a payload of `1.6*4/8=0.8GB` per GPU - so now we will look at the output of nccl-tests and match the throughput report for ~0.8GB payload - is that good?

> I normally think of the factors that matter for all-gather time as the all-gather output size and the world size. (The all-gather input size can be derived from those two by dividing.) The all-gather output size affects the bandwidth term, and the world size affects the latency term. When you define the comms. buffer size, I am curious why you define it using the all-gather input size (with the `/8` factor)?

Because if you look at the actual benchmark code - the reported payload in the nccl-tests is per GPU and not per world size. say, in the benchmark each rank all-gathers 4GB of payload (so for 8 gpus it'll be 32GB total), and one will need to look up payload of 4GB and not 32GB in the report.

Does that explain why or am I misunderstanding something?

Here is the pure python implementation of the all-reduce benchmark w/ a hardcoded 4GB payload per rank.
https://github.com/stas00/ml-engineering/blob/master/multi-node/all_reduce_bench.py
its results correspond to nccl-tests's record for 4GB! So that how I knew it's per rank and not per world size.

Reply:
@stas00 Thanks for the clarification! I was not familiar with the nccl-tests, but what you said makes sense!

Reply:
ok, I compiled your awesome design notes into a doc, adding my numerical examples, and splitting the sections into clear sub-topics so hopefully the reader can follow things easily.

I know it has grown a lot since the original doc and you didn't really want to spend time on it, so I apologize again for dragging you into doing some of the work anyway, Andrew.

I propose that as long as I didn't add anything incorrect we can just merge it as is and have future PRs to improve upon this already amazing practical information you shared. The original intention was just to preserve your notes. and for those of us who need to do the math to be able to do it intelligently which these notes enable.